content/blog/2019-10-27-tbjit-type-specialization/index.md

```diff
@@ -0,0 +1,189 @@
+
+<img src="state-machine.png" alt="Interpreter Performance" width="100%">
+
+Machine code emitted by a JIT might look something like the following. Notice that there are fewer 'overhead' instructions than in the interpreter version (just two instead of seven).
+
+<img src="guard-vvadd.png" alt="Interpreter Performance" width="40%">
+
+## TraceMonkey
+
+*TraceMonkey* is the JIT proposed by the authors and roughly follows the high-level ideas described above. The main steps of TraceMonkey are *interpreting*, *recording*, *compilation*, *native execution*, and *aborting*.
+
+### Interpreting
+
+The default state of TraceMonkey is to execute bytecode via an interpreter. This yields correct, but slow execution of a users program.
+
+### Recording
+
+When TraceMonkey detects a loop (simply a back-edge in the control-flow graph), it begins to record a trace. For each bytecode instruction, one or more *LIR* instructions are generated along with type guards. *LIR* instructions directly map to machine instructions, but are ISA agnostic.
+
+```c
+// Bytecode (not typed)
+c = add a b;
+
+// LIR Trace (typed)
+guard typeof(a) == int;
+guard typeof(b) == int; 
+int c = add_int a b;
+```
+
+Traces inherently can only follow a single path within a loop iteration. No type information is known about the paths that we not taken, so we can't generate machine code for them. Therefore guards must also check branch conditions. In the following example, two possible traces can be generated from the code.
+
+```c
+// Source code
+if (a == true) {
+  c++;
+}
+else {
+  c--;
+}
+
+// LIR Trace 1
+guard a == true;
+int c = add_int c 1;
+
+// LIR Trace 2
+guard a == false;
+int c = add_int c -1;
+
+```
+
+To avoid confusion, keep in mind that there are four types of code in TraceMonkey. 1) Source code, 2) Bytecode, 3) LIR, and 4) Machine code. Only bytecode and machine code are executed, while source code and LIR are only meant to compiled down to the subsequent code level.
+
+### Compilation
+
+The LIR traces must first be compiled to machine code to execute natively on the processor. This compilation needs to be much faster than static compilation because it occurs during runtime. The authors propose limiting the number of code optimizations performed to keep the compilation runtime reasonable. For example, register allocation uses a greedy algorithm. Although the authors did not evaluate this, greedy algorithms generally give non-optimal results, but have an excellent accuracy (compared to oracle) per time ratio.
+
+The compiled traces are stored in a trace buffer for later use by the interpreter.
+
+```c
+// LIR code
+guard a == true;
+int c = add_int c 1;
+
+// Machine code
+addi t0 x0 1;
+bne  t1 t0 abort;
+addi t2 t2 1;
```

Comment:
Not quite sure I see why there are two `addi`s here. Is it part of the type tag check?